presto-raptor/src/test/java/com/facebook/presto/raptor/storage/TestBucketBalancer.java

```diff
@@ -0,0 +1,349 @@
+import com.facebook.presto.raptor.metadata.BucketNode;
+import com.facebook.presto.raptor.metadata.ColumnInfo;
+import com.facebook.presto.raptor.metadata.Distribution;
+import com.facebook.presto.raptor.metadata.MetadataDao;
+import com.facebook.presto.raptor.metadata.ShardManager;
+import com.facebook.presto.spi.Node;
+import com.facebook.presto.testing.TestingNodeManager;
+import com.facebook.presto.type.TypeRegistry;
+import com.google.common.collect.HashMultiset;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Multiset;
+import io.airlift.units.DataSize;
+import io.airlift.units.Duration;
+import org.skife.jdbi.v2.DBI;
+import org.skife.jdbi.v2.Handle;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import java.net.URI;
+import java.util.List;
+import java.util.OptionalLong;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.raptor.metadata.Distribution.serializeColumnTypes;
+import static com.facebook.presto.raptor.metadata.SchemaDaoUtil.createTablesWithRetry;
+import static com.facebook.presto.raptor.metadata.TestDatabaseShardManager.createShardManager;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.airlift.testing.Assertions.assertGreaterThanOrEqual;
+import static io.airlift.testing.Assertions.assertLessThanOrEqual;
+import static java.util.concurrent.TimeUnit.DAYS;
+import static org.testng.Assert.assertEquals;
+
+@Test(singleThreaded = true)
+public class TestBucketBalancer
+{
+    private static final List<String> AVAILABLE_WORKERS = ImmutableList.of("node1", "node2", "node3", "node4", "node5");
+
+    private DBI dbi;
+    private Handle dummyHandle;
+    private ShardManager shardManager;
+    private TestingNodeManager nodeManager;
+    private MetadataDao metadataDao;
+    private BucketBalancer balancer;
+
+    @BeforeMethod
+    public void setup()
+            throws Exception
+    {
+        TypeRegistry typeRegistry = new TypeRegistry();
+        dbi = new DBI("jdbc:h2:mem:test" + System.nanoTime());
+        dbi.registerMapper(new Distribution.Mapper(typeRegistry));
+        dummyHandle = dbi.open();
+        createTablesWithRetry(dbi);
+
+        metadataDao = dbi.onDemand(MetadataDao.class);
+        nodeManager = new TestingNodeManager(AVAILABLE_WORKERS.stream()
+                .map(TestBucketBalancer::createTestingNode)
+                .collect(Collectors.toList()));
+
+        NodeSupplier nodeSupplier = nodeManager::getWorkerNodes;
+        shardManager = createShardManager(dbi, nodeSupplier);
+        balancer = new BucketBalancer(nodeSupplier, shardManager, new Duration(1, DAYS), true, true, "test");
+    }
+
+    @AfterMethod(alwaysRun = true)
+    public void teardown()
+    {
+        if (dummyHandle != null) {
+            dummyHandle.close();
+        }
+    }
+
```

Comment:
Can you add some tests with an *odd* bucket count (like 17) and both less and more buckets than the number of nodes?